src/lndk_offers.rs

```diff
@@ -88,28 +90,30 @@ impl Display for OfferError<Secp256k1Error> {
     }
 }
 
-impl Error for OfferError<Secp256k1Error> {}
+impl Error for OfferError {}
 
 // Decodes a bech32 string into an LDK offer.
 pub fn decode(offer_str: String) -> Result<Offer, Bolt12ParseError> {
     offer_str.parse::<Offer>()
 }
 
 impl OfferHandler {
-    pub async fn send_invoice_request(
-        &self,
-        mut cfg: PayOfferParams,
-    ) -> Result<u64, OfferError<bitcoin::secp256k1::Error>> {
+    pub async fn send_invoice_request(&self, mut cfg: PayOfferParams) -> Result<u64, OfferError> {
         let validated_amount = validate_amount(&cfg.offer, cfg.amount).await?;
 
         // For now we connect directly to the introduction node of the blinded path so we don't need
         // any intermediate nodes here. In the future we'll query for a full path to the
         // introduction node for better sender privacy.
         match cfg.destination {
             Destination::Node(pubkey) => connect_to_peer(cfg.client.clone(), pubkey).await?,
-            Destination::BlindedPath(ref path) => {
-                connect_to_peer(cfg.client.clone(), path.introduction_node_id).await?
-            }
+            Destination::BlindedPath(ref path) => match path.introduction_node {
+                IntroductionNode::NodeId(pubkey) => {
+                    connect_to_peer(cfg.client.clone(), pubkey).await?
+                }
+                IntroductionNode::DirectedShortChannelId(..) => {
+                    unimplemented!()
```

Comment:
Fixing this `unimplemented` and the next one in a different file

Reply:
so in order to get the pubkey in the `DirectedShortChannelId` scenario we would need to access the network graph(could be `ReadOnlyNetworkGraph`  or `NetworkGraph`) in order to get the channel by the `short_channel_id` returned in the second argument from `DirectedShortChannelId` and then extract the pubkey with the help of the first argument, `Direction`. This would require to intialise the network graph inside of `main.rs` before initialising the `OfferHandler` and `LndkOnionMessenger` and probably others.

Reply:
Oh sure, we don't have LDK's NetworkGraph implemented because it's not 100% needed quite yet. I think we can use LND's API to look up this info for now: https://lightning.engineering/api-docs/api/lnd/lightning/get-chan-info

Reply:
Ah yup that would be great! And note that when calling LND's api throughout LNDK we put the the calls behind traits: https://github.com/lndk-org/lndk/blob/d1c668cb2b8e4a1b3ed78a72e125859278c2f49a/src/lnd.rs#L333 This is mainly to make mocking the LND api in the unit tests easier. Let me know if you have any questions or if I can help there

Reply:
Been looking into getting the channel info from lnd and try to use it where needed but had some challenges as `lnd` expects `&mut self` while the `NodeIdLookUp` trait from `ldk` expects `&self`  and also `lnd` methods are async and calling them in `next_node_id` function from the `NodeIdLookUp` trait would require to use something like `tokio::...block_on`.

Let me know if you have any thoughts.

Here is what I roughly tried

Added the following trait:
```rust
#[async_trait]
pub trait ChannelManager {
    async fn channel_info(&mut self, scid: u64) -> Result<ChannelEdge, Status>;
}
```
then implemented for `Client`:
```rust
#[async_trait]
impl ChannelManager for Client {
    async fn channel_info(&mut self, scid: u64) -> Result<ChannelEdge, Status> {
        let chan_info_req = ChanInfoRequest {
            chan_id: scid,
        };
        self.lightning()
            .get_chan_info(chan_info_req)
            .await
            .map(|resp| resp.into_inner())
    }
}
```
I want to implement the `NodeIdLookUp` trait from `LDK` as it required for the `OnionMessenger` and also can be used to resolve the `unimplemented` code in case of `IntroductionDirectedShortChannelId`:
```rust
pub struct LndkNodeIdLookUp {
    client: Client,
    our_node_id: PublicKey,
}

impl LndkNodeIdLookUp {
    pub fn new(client: Client, our_node_id: PublicKey) -> Self {
        LndkNodeIdLookUp {
            client,
            our_node_id,
        }
    }
}

impl NodeIdLookUp for LndkNodeIdLookUp {
    fn next_node_id(&self, scid: u64) -> Option<PublicKey> {
        let chan_info = self.client.channel_info(scid); <-- this is async and also requires &mut
        // extract counterparty node id from chan_info
        // return Some(counterparty_node_id)
        None
    }
}
```

Reply:
Ah yes, we've hit these annoying problem before unfortunately. :( 

To solve the `&mut self` vs `&self` trait problem, iirc we ended up using a RefCell before, but I'd rather not do that again... I'm wondering if we can add a change upstream to tonic_lnd to add a "readonly" version of the client. I'll give it a try sometime today and get back to you.

Reply:
For the &mut self vs &self trait problem, I think this branch should work if you can change the reference to tonic_lnd in Cargo.toml :) https://github.com/orbitalturtle/tonic_lnd/pull/4

You can call `client.lightning_read_only()` instead of `client.lightning()` to get the client. I tried running lndk with the lightning_read_only() client and it seemed to work as expected. I didn't try calling get_chan_info in particular. I assume that will work too... but let me know. 

That doesn't solve the async/sync discrepancy... I think tokio block_on should be ok for now